src/pages/audio-rewards-page/hooks.ts

```diff
@@ -0,0 +1,21 @@
+import { useSelector } from 'react-redux'
+
+import { ChallengeRewardID } from 'common/models/AudioRewards'
+import { getCompletionStages } from 'components/profile-progress/store/selectors'
+
+type OptimisticChallengeCompletionResponse = Partial<
+  Record<ChallengeRewardID, number>
+>
+
+export const useOptimisticChallengeCompletionStepCounts = () => {
```

Comment:
Can this take a challenge object and return a challenge object with the step count updated as appropriate? Then we don't have to have the repeated override logic.

We could also shorten the name to `useOptimisticChallenge` then.

Something like:


```ts
export const useOptimisticChallenge(challenge: UserChallenge) => {
  const optimisticSteps = challengeStepCalculators[challenge.id]() // internal map of challenge step calculators
  if (optimisticSteps) {
      return { ...challenge, step_count: optimisticSteps }
  }
  return challenge
}
```